lib/rules/indent.js

```diff
@@ -415,9 +436,10 @@ class OffsetStorage {
 
             if (this._ignoredTokens.has(token)) {
 
-                // If the token is ignored, use the actual indent of the token as the desired indent.
-                // This ensures that no errors are reported for this token.
-                this._desiredIndentCache.set(token, this._tokenInfo.getTokenIndent(token).length / this._indentSize);
+                // If the token is ignored, set the desired indent to an IgnoredTokenIndent
+                const observedIndent = this._tokenInfo.getTokenIndent(token).length / this._indentSize;
+
+                this._desiredIndentCache.set(token, new IgnoredTokenIndent(observedIndent));
```

Comment:
I tried something like that at first, and it resulted in a large number of test failures. So I don’t think this will work.

The reason is – as I understand it, which may be imperfectly – that the indentation of an ignored line may nevertheless influence the expected indentation of subsequent lines. In code terms, the value is used [here](https://github.com/eslint/eslint/blob/ee99876ccc76585b0af2c8673d7eb25bb0b972fa/lib/rules/indent.js#L441).

Perhaps someone who understands this better can give a clearer explanation; but in any case having unsuccessfully tried something like this I don’t think it’s a goer, sadly.

Reply:
@Aladdin-ADD You can see the sort of problems your proposal would cause by changing the `valueOf()` method of `class IgnoredTokenIndent` to return 0 rather than `this._indent`.

```diff
diff --git a/lib/rules/indent.js b/lib/rules/indent.js
index 3c22ad0..936d1d0 100644
--- a/lib/rules/indent.js
+++ b/lib/rules/indent.js
@@ -243,7 +243,7 @@ class IgnoredTokenIndent {
      * @returns {number} The indent
      */
     valueOf() {
-        return this._indent;
+        return 0; //this._indent;
     }
 }
 
```

Here’s what I get when I run the tests with that change:

```
$ npm test

> eslint@4.8.0 test /Users/robin/code/eslint
> node Makefile.js test

Validating Makefile.js
Validating JSON Files
Validating Markdown Files
Validating JavaScript files

/Users/robin/code/eslint/lib/ast-utils.js
  539:1  error  Expected indentation of 4 spaces but found 16  indent
  542:1  error  Expected indentation of 0 spaces but found 12  indent

/Users/robin/code/eslint/lib/code-path-analysis/code-path-state.js
  474:1  error  Expected indentation of 0 spaces but found 16  indent

/Users/robin/code/eslint/lib/rules/array-bracket-newline.js
  194:1  error  Expected indentation of 4 spaces but found 20  indent
  197:1  error  Expected indentation of 0 spaces but found 16  indent

/Users/robin/code/eslint/lib/rules/array-bracket-spacing.js
  193:1  error  Expected indentation of 4 spaces but found 20  indent
  199:1  error  Expected indentation of 4 spaces but found 20  indent

/Users/robin/code/eslint/lib/rules/array-element-newline.js
  193:1  error  Expected indentation of 4 spaces but found 20  indent
  195:1  error  Expected indentation of 0 spaces but found 16  indent

/Users/robin/code/eslint/lib/rules/indent.js
  245:12  error  Expected 'this' to be used by class method 'valueOf'          class-methods-use-this
  246:19  error  Expected exception block, space or tab after '//' in comment  spaced-comment

/Users/robin/code/eslint/lib/rules/keyword-spacing.js
  495:1  error  Expected indentation of 4 spaces but found 20  indent
  497:1  error  Expected indentation of 0 spaces but found 16  indent

/Users/robin/code/eslint/lib/rules/no-cond-assign.js
  92:1  error  Expected indentation of 4 spaces but found 20  indent
  93:1  error  Expected indentation of 4 spaces but found 20  indent
  94:1  error  Expected indentation of 0 spaces but found 16  indent

/Users/robin/code/eslint/lib/rules/no-extra-parens.js
  375:1  error  Expected indentation of 4 spaces but found 24  indent
  376:1  error  Expected indentation of 4 spaces but found 24  indent
  377:1  error  Expected indentation of 4 spaces but found 24  indent
  379:1  error  Expected indentation of 0 spaces but found 20  indent
  467:1  error  Expected indentation of 4 spaces but found 20  indent
  469:1  error  Expected indentation of 4 spaces but found 24  indent
  472:1  error  Expected indentation of 0 spaces but found 20  indent
  474:1  error  Expected indentation of 0 spaces but found 16  indent
  570:1  error  Expected indentation of 4 spaces but found 32  indent
  571:1  error  Expected indentation of 0 spaces but found 28  indent
  614:1  error  Expected indentation of 4 spaces but found 24  indent
  616:1  error  Expected indentation of 4 spaces but found 28  indent
  620:1  error  Expected indentation of 0 spaces but found 24  indent
  621:1  error  Expected indentation of 0 spaces but found 20  indent

/Users/robin/code/eslint/lib/rules/no-implicit-coercion.js
  81:1  error  Expected indentation of 4 spaces but found 12  indent
  84:1  error  Expected indentation of 0 spaces but found 8   indent

/Users/robin/code/eslint/lib/rules/no-lonely-if.js
  57:1  error  Expected indentation of 4 spaces but found 36  indent
  61:1  error  Expected indentation of 0 spaces but found 32  indent

/Users/robin/code/eslint/lib/rules/no-mixed-operators.js
  121:1  error  Expected indentation of 4 spaces but found 20  indent
  123:1  error  Expected indentation of 0 spaces but found 16  indent

/Users/robin/code/eslint/lib/rules/no-multi-spaces.js
  92:1  error  Expected indentation of 4 spaces but found 28  indent
  94:1  error  Expected indentation of 0 spaces but found 24  indent

/Users/robin/code/eslint/lib/rules/no-unused-vars.js
  385:1  error  Expected indentation of 4 spaces but found 20  indent
  387:1  error  Expected indentation of 0 spaces but found 16  indent
  391:1  error  Expected indentation of 4 spaces but found 20  indent
  394:1  error  Expected indentation of 0 spaces but found 16  indent

/Users/robin/code/eslint/lib/rules/no-useless-call.js
  25:1  error  Expected indentation of 4 spaces but found 12  indent
  27:1  error  Expected indentation of 0 spaces but found 8   indent

/Users/robin/code/eslint/lib/rules/no-useless-constructor.js
  133:1  error  Expected indentation of 4 spaces but found 12  indent
  135:1  error  Expected indentation of 0 spaces but found 8   indent

/Users/robin/code/eslint/lib/rules/object-curly-newline.js
  141:1  error  Expected indentation of 4 spaces but found 20  indent
  144:1  error  Expected indentation of 0 spaces but found 16  indent

/Users/robin/code/eslint/lib/rules/padding-line-between-statements.js
  116:1  error  Expected indentation of 4 spaces but found 12  indent
  118:1  error  Expected indentation of 4 spaces but found 16  indent
  120:1  error  Expected indentation of 0 spaces but found 12  indent
  121:1  error  Expected indentation of 0 spaces but found 8   indent

/Users/robin/code/eslint/lib/rules/prefer-const.js
  51:1  error  Expected indentation of 4 spaces but found 12  indent
  54:1  error  Expected indentation of 0 spaces but found 8   indent

/Users/robin/code/eslint/lib/rules/space-before-function-paren.js
  73:1  error  Expected indentation of 4 spaces but found 24  indent
  76:1  error  Expected indentation of 0 spaces but found 20  indent
  77:1  error  Expected indentation of 0 spaces but found 16  indent

/Users/robin/code/eslint/lib/rules/vars-on-top.js
  59:1  error  Expected indentation of 4 spaces but found 20  indent
  62:1  error  Expected indentation of 0 spaces but found 16  indent

✖ 59 problems (59 errors, 0 warnings)
  58 errors, 0 warnings potentially fixable with the `--fix` option.

/Users/robin/code/eslint/node_modules/shelljs/src/common.js:417
      if (config.fatal) throw e;
                        ^

Error: exec: internal error
    at Object.error (/Users/robin/code/eslint/node_modules/shelljs/src/common.js:128:27)
    at _exec (/Users/robin/code/eslint/node_modules/shelljs/src/exec.js:292:12)
    at /Users/robin/code/eslint/node_modules/shelljs/src/common.js:350:23
    at Function.target.lint (/Users/robin/code/eslint/Makefile.js:532:18)
    at Object.global.target.(anonymous function) [as lint] (/Users/robin/code/eslint/node_modules/shelljs/make.js:36:40)
    at Function.target.test (/Users/robin/code/eslint/Makefile.js:585:12)
    at Object.global.target.(anonymous function) [as test] (/Users/robin/code/eslint/node_modules/shelljs/make.js:36:40)
    at /Users/robin/code/eslint/node_modules/shelljs/make.js:48:27
    at Array.forEach (native)
    at Timeout._onTimeout (/Users/robin/code/eslint/node_modules/shelljs/make.js:46:10)
npm ERR! Test failed.  See above for more details.
```

Reply:
yes, you are right! this is more complicated than I thought. ~ovvo~

But I'm afraid it's not a good idea to store a object like this.maybe there is something wrong `this._tokenInfo.getTokenIndent(token)`, we need to figure out.

Reply:
I don’t think there’s anything wrong with `getTokenIndent(token)`, which seems to be working as intended.

I guess it’s a matter of taste whether one considers the use of a marker class to be strange. I’m certainly open to alternative implementations if you have any suggestions. The only thing I really care about is getting this fixed somehow.

Reply:
(I’ve added a link to an alternative implementation in a top-level comment below.)